### ISSUE TITLE
How I can stop blink of cursor in Windows Terminal?

### ISSUE THREAD

**Issue by KiYugadgeter** [microsoft/terminal#1379]
I don't link blinking cursor on terminal.
so I have tried to stop blink of cursor from settings but there is no option to do it.
How to disable blink of cursor?

**Comment by leptr** [comment]
I just set the cursor color to match the background color so it would be invisible, I couldn't find any other way to either remove it or stop it from blinking, but I hope this helps you

**Comment by Lo0oG** [comment]
Go into the Control Panel and search for Blink.

**Comment by AeliusSaionji** [comment]
Seconding a request for this setting.

**Comment by leptr** [comment]
also I would love to see an option to make it look like that box cursor that we have in linux terminal, I think that would be a nice feature

**Comment by DHowett-MSFT** [comment]
What, you mean this?
![image](https://user-images.githubusercontent.com/14316954/60136920-5e82ec00-975a-11e9-9be1-07db240bccd3.png)


Set your cursor to `filledBox`.

**Comment by leptr** [comment]
> What, you mean this?
> ![image](https://user-images.githubusercontent.com/14316954/60136920-5e82ec00-975a-11e9-9be1-07db240bccd3.png)
> 
> Set your cursor to `filledBox`.

Hey, thanks! I didn't know the different cursor names

**Comment by KiYugadgeter** [comment]
You could see here.
https://github.com/microsoft/terminal/issues/1294#issuecomment-502539208

**Comment by AeliusSaionji** [comment]
Or reference the nice documentation here https://github.com/microsoft/terminal/blob/master/doc/cascadia/SettingsSchema.md

**Comment by dkter** [comment]
The cursor blinking can be turned off (system-wide) by going to Keyboard Properties in the Control Panel and turning the cursor blink rate all the way down:

![image](https://user-images.githubusercontent.com/34610081/60402879-67612c80-9b63-11e9-9b5a-00ab549f2327.png)

Would it be useful to disable the cursor blinking for just the terminal app?

**Comment by KiYugadgeter** [comment]
I want to disable it in terminal app only

2019年7月1日(月) 7:19 David Teresi <notifications@github.com>:

**Comment by SCWells72** [comment]
Agreed that this should be a distinct option just for the terminal app. Here's how it can be configured in mintty:

![image](https://user-images.githubusercontent.com/7671043/60527999-0f3c3e80-9cb9-11e9-9ea6-e4c9e67acae8.png)

Note the distinct config fields for cursor shape and whether to blink.

**Comment by KiYugadgeter** [comment]
I've tried write some code to disable it.
But it looks like there are some error. (IDE has indicated redefinition of method  definition of UseBlinkCursor in TerminalSettings.cpp despite I have defined just one time.)

https://github.com/KiYugadgeter/terminal/tree/my_build

**Comment by hwo411** [comment]
Any ETA to add this?

**Comment by zadjii-msft** [comment]
@hwo411 Nope - It's marked as "Terminal Backlog", which means we don't think we're going to get to it ourselves before we release a 1.0 version of the terminal.

It's also marked "Help-Wanted", so if someone's feeling particularly passionate about this, they should feel free to build a solution themselves, and we'd happily review a PR 😉

**Comment by hwo411** [comment]
@zadjii-msft thanks for explanation!

**Comment by Kagami** [comment]
@skyline75489 created PR for this feature: https://github.com/microsoft/terminal/pull/2892

**Comment by qqkookie** [comment]
Blinking is only binary on/off option, so how about merging blinking attribute and cursor shape attribute into one attribute, like "filledbox" vs "static-filledbox", or "blinking-filledbox" vs "filledbox"?

The name "solid" for non-blinking cursor is confusing. Normally "solid box" means "filledbox". So "solid-emptybox" may be confusing. I prefer "static" as non-blinking attribute name.

The cursor name "filledbox" is rather strange too.  I suggest to use name like "solidbox" or "fullbox" (Unicode name) for current "filledbox" attribute value name. 

The cursor name "vintage" is also not intuitive. How about " quarterblock"? 

Most people will use just one of "static-solidbox" or "blinking-bar". IMHO, default should be static  and solidbox (filledbox).  

And if wt window lose input focus, cursor should stop blinking and  turned into static-emptybox.

**Comment by sirfz** [comment]
Possibly relevant: is it possible to make the filledBox cursor transparent? Right now (at least in my case) the box is opaque while emptyBox just doesn't have the same feel to what I'm used to (a block which still displays what's under it).

How it looks in Windows Terminal:

![image](https://user-images.githubusercontent.com/4741099/74553723-98f6fe80-4f60-11ea-88c6-1caf25b937a3.png)

How it looks in other terminals (in this case alacritty):

![image](https://user-images.githubusercontent.com/4741099/74553674-8250a780-4f60-11ea-987e-b52b34861b71.png)

**Comment by AndersBillLinden** [comment]
It seems like the cursor blink rate setting in control panel only affects the terminal window. Other applications still blinks the cursor.

**Comment by Avi0** [comment]
> Go into the Control Panel and search for Blink.

Disabling cursor blinking system wide causes Windows taskbar icon notifications not blink and not disappear. This broken Windows functionality prevents users from disabling cursor blinking. Since this issue exists for years, nobody is going to fix it in Windows. Hence, unfortunately, cursor blink rates should be controlled by applications.

Lack of this feature in Windows Terminal is the only thing that prevents me from switching to Windows Terminal completely.

**Comment by mdtauk** [comment]
If blinking becomes a cursor type or state, could there be an option for both a binary on and off, as well as an animated fade for the cursor?

**Comment by dkter** [comment]
@mdtauk could you maybe point to an example of a program with an animated fading cursor? I'm not sure what exactly you're proposing.

**Comment by mdtauk** [comment]
I can't, but it would look smoother, and I think would work well with the vintage terminal visual effects. It only needs to be a few frames of the cursor fading in and then out again

**Comment by thegreyshadow** [comment]
I second this request. Please add a way to turn off blinking cursor in terminal without resorting to any system-wide hack. Thanks in advance, I would really like to have this feature added to this great app.

**Comment by Avi0** [comment]
> without resorting to any system-wide hack.

The problem is that there is no acceptable system-wide hack to resort to. If you disable cursor blink globally, you'll end up with broken taskbar icon notifications. Icons will glow orange permanently.

**Comment by DHowett** [comment]
> permanently

"Until you interact with them, clearing their urgency state."

**Comment by thegreyshadow** [comment]
> > without resorting to any system-wide hack.
> 
> The problem is that there is no acceptable system-wide hack to resort to. If you disable cursor blink globally, you'll end up with broken taskbar icon notifications. Icons will glow orange permanently.

My point exactly.

**Comment by Avi0** [comment]
> > permanently
> 
> "Until you interact with them, clearing their urgency state."

No! This is the annoyance exactly. The state is not cleared after interaction, they stay orange. I found a workaround, switch to another desktop and back, but this is still an annoyance.

**Comment by Avi0** [comment]
Looks like there is a system-wide workaround. I've been using it for a few days and I haven't experienced any problems so far.

1. Disable cursor blinking in Keyboard settings (`control main.cpl keyboard`).
2. In Advanced System Settings (`SystemPropertiesAdvanced`), under Performance, deselect "Animations in the taskbar".

Cursor blinking will be disabled system-wide without adverse side effects mentioned above.

**Comment by igor-vivaldi** [comment]
The original reason for a blinking cursor was that a thin vertical line was not very noticeable. But this is not applicable to a solid filled box with inverse video. It is very visible. Adding blinking makes it too visible to the point of distraction. For this reason a system settings that assumes a thin line should not apply when the cursor is a filled box. 

This suggest that blinking or not should be a part of cursor shape.

**Comment by mangelozzi** [comment]
> The original reason for a blinking cursor was that a thin vertical line was not very noticeable. But this is not applicable to a solid filled box with inverse video. It is very visible. Adding blinking makes it too visible to the point of distraction. For this reason a system settings that assumes a thin line should not apply when the cursor is a filled box.
> 
> This suggest that blinking or not should be a part of cursor shape.

I agree with you, but not everyone is the same (personality differents), someone may want a flashing big box and think its perfect. An independat settings would mean everyone is then happy. Unfortunately WT does not support inverse video yet, so this is not the case (there are time when white cursor is on light text and I can barely see it).

**Comment by wusticality** [comment]
What is the status on this? It's driving me nuts that the cursor blinks within `Emacs` :(

**Comment by yan12125** [comment]
Printing a DECSCUSR command from bash/zsh stops the cursor from blinking:
```
echo -e "\e[2 q"
```
More options can be found at https://github.com/MicrosoftDocs/Console-Docs/issues/140.

I don't know how to achieve it in cmd or powershell, though.

**Comment by thenger** [comment]
in powershell you can do:

```powershell
"`e[2 q"
```

or with python:
```
python -c "print('\033[2 q')"
```

**Comment by ndavidpsi** [comment]
Any updates on how to have a non blinking cursor?

**Comment by stanio** [comment]
> Any updates on how to have a non blinking cursor?

Per previous comments, I've created a `cmd` (`steady_cur.cmd`) script on my `PATH`:

```cmd
@echo <esc>[2 q
```

Where `<esc>` is a single ASCII character `0x1B` I've entered into the file using a hex-editor.  Appears to work with all of Command Prompt, PowerShell and Git Bash.  You could change the `"commandline"` of the Command Prompt profile in the Windows Terminal Settings (`settings.json`) to:

```json
                "commandline": "cmd.exe /k steady_cur",
```

to have it automatically applied in new consoles.

**Comment by ndavidpsi** [comment]
how nice, I'll try it

**Comment by ndavidpsi** [comment]
Alas it did not work, is the syntax correct?

**Comment by stanio** [comment]
```cmd
@echo <esc>[2 q
```
You need to replace `<esc>` with the byte `0x1B`, and not the literal text `0x1B` (not sure what you're actually trying).

**Comment by ndavidpsi** [comment]
I did this exactly `@echo <esc>[2 q`

**Comment by ndavidpsi** [comment]
Update: sorry I didn't read it properly before. It works great now!

**Comment by avdi** [comment]
Pleeeeeeeeeese 🙏

I make screencasts and I normally disable cursor blink in my editors and terminals so that I can speed up, slow down, and freeze video without distracting changes in cursor blink rate.

**Comment by rivy** [comment]
For ease-of-use at CMD...

```batch
@rem cursor.cmd ~ `cursor [[block | underline | bar] [blink]"`
@setLocal
@echo off

:: set cursor style (via ANSI escape sequence)
:: * ref: [Cursor Escape Sequences](https://github.com/MicrosoftDocs/Console-Docs/issues/140)
:: * ref: https://github.com/microsoft/terminal/issues/1379#issuecomment-735679639

set "seq=[0 q" &::rem "default cursor"
if /i "%~1"=="block" set "seq=[2 q"
if /i "%~1"=="underline" set "seq=[4 q"
if /i "%~1"=="bar" set "seq=[6 q"
if /i "%~1"=="block" if /i "%~2"=="blink" set "seq=[1 q"
if /i "%~1"=="underline" if /i "%~2"=="blink" set "seq=[3 q"
if /i "%~1"=="bar" if /i "%~2"=="blink" set "seq=[5 q"

powershell -noninteractive -noprofile -command [char]27+'%seq%'
```

**Comment by strawhat-dev** [comment]
For those wondering, the easiest way I've found to disable the blinking cursor is by putting `echo -e -n "\e[2 q"` at the end of your .bashrc/.zshrc file (or ```Write-Host -NoNewLine "`e[2 q"``` at the end of your profile.ps1 for PowerShell users). This way, they get called upon startup. As a bonus, the -n/-NoNewLine flag removes the extra new line that appears on startup which bugged me a little.

**Comment by rivy** [comment]
You can alternatively use non-standard command lines to start a given shell, such as `"commandline": "cmd.exe /k cursor bar"`.

**Comment by fweep** [comment]
> This suggest that blinking or not should be a part of cursor shape.

Please not this. I very much want a bar cursor that doesn't blink.

**Comment by LoganDark** [comment]
Still waiting on this. A cursor blink is extremely obnoxious while looking at the terminal and trying to read text, Windows Terminal is still extremely limited in its customization options and basic features like this should _not_ be backlogged. Please reconsider your stance on this, cursor blink in text boxes has nothing to do with cursor blink in terminals, so my terminal blink should not be controlled by that global setting.

**Comment by divijan** [comment]
> 
> 
> For those wondering, the easiest way I've found to disable the blinking cursor is by putting `echo -e -n "\e[2 q"` at the end of your .bashrc/.zshrc file (or `` Write-Host -NoNewLine "`e[2 q" `` at the end of your profile.ps1 for PowerShell users). This way, they get called upon startup. As a bonus, the -n/-NoNewLine flag removes the extra new line that appears on startup which bugged me a little.

worked like charm on Ubuntu WSL2 👍 Thanks

**Comment by eddiechaoinchina** [comment]
Is there any workaround here without changing my shell startup file? I tried it and it smh broke some of my vim plugins. PLZ make it a customizable options for WT.😢

**Comment by zeroxia** [comment]
> > without resorting to any system-wide hack.
> 
> The problem is that there is no acceptable system-wide hack to resort to. If you disable cursor blink globally, you'll end up with broken taskbar icon notifications. Icons will glow orange permanently.

Microsoft can integrate Linux because it's freshly new project. But it's really hard for them to cooperate internally among different teams to fix some stupidly trivial but tremendously annoying glitches like this. This problem was reported years ago and I just have it on Windows 10 20H2 build 19042.

https://superuser.com/questions/1114598/windows-10-taskbar-icons-dont-stop-glowing-orange

https://www.reddit.com/r/Windows10/comments/4qlgqp/taskbar_icons_stay_highlighted_orange_forever/

**Comment by Avi0** [comment]
The workaround is to disable cursor blinking system-wide and disable taskbar animations in System Properties, Advanced, Visual Effects (to avoid stuck icon highlighting in the taskbar).

**Comment by willchao612** [comment]
It's weird that I have done nothing and it just stopped blinking anymore since today.

**Comment by willchao612** [comment]
It's weird that I have done nothing and it just started blinking again since today.

**Comment by LoganDark** [comment]
@willchao612 A story in two comments...

**Comment by willchao612** [comment]
Finally got a solution. If you are using tmux and this cursor blinking issue happens, put the following line in your `.tmux.conf` and restart your tmux session.

```tmux
set -sa terminal-overrides '*:Ss=\E[%p1%d q:Se=\E[ q'
```

Found this by chance when solving another cursor shape issue in Neovim. [Docs](https://neovim.io/doc/user/term.html#tui-cursor-tmux) here. I hope it helps you guys.

**Comment by zakutin** [comment]
> 
> 
> It's weird that I have done nothing and it just stopped blinking anymore since today.

I noticed the following:
* launch wt, cursor is blinking
* run neovim (by default the cursor is not blinking in neovim)
* exit neovim
* cursor is _not blinking_ in wt command line
* restart wt, cursor is blinking again

**Comment by willchao612** [comment]
@zakutin Did you manually set TERM in you startup file? If so, try commenting it out and restart tmux.

**Comment by zadjii-msft** [comment]
FWIW This probably got better in some ways in #10543/#10589, and worse in others: #10754

**Comment by ZYinMD** [comment]
Hey guys, maybe a dumb question but: does the cursor in Linux blink by default? 

I use the cursor shape to remind myself where I am - in cmd and pwsh I use a blinking underscore, in git bash I use a blinking pipe, in WSL I use a block. Those are their original default shapes, so I try to keep them. 

So, now the question is, if I want my Ubuntu to look authentic, should I make the block blink or not blink?

**Comment by niask11** [comment]
> Hey guys, maybe a dumb question but: does the cursor in Linux blink by default?

In Ubuntu Desktop it's a block and it doesn't blink as far as I remember...

**Comment by dkter** [comment]
@ZYinMD in GNOME Terminal (the default terminal in Ubuntu) the cursor is a blinking solid block, in the TTY (i.e. if you run Ubuntu without a GUI) it's a blinking underscore

**Comment by mataha** [comment]
Hi! For those subscribed to this thread and many more coming from search results like me, I have a `cmd.exe` solution ready.

<details>
<summary>The code</summary>

```batchfile
@setlocal EnableDelayedExpansion EnableExtensions & set "CMD=!CMDCMDLINE!" 2>nul
@if /i not "!CMD!"=="!CMD:/=!" (goto :EOF) else @if not defined DEBUG (echo off)

for /f "usebackq" %%E in (`echo(prompt $E^| "%ComSpec%" /d /q "" 2^>nul`) do (
    echo(%%~E[?12l %%~E[3A %%~E[3M %%~E[1A
)
```

Save the above to `~\.cmd` as UTF-8, then run:

```batchfile
set PROCESSOR_ARCHITECTURE | "%SystemRoot%\system32\find.exe" "64" >nul && (
    set "registry_view=64") || set "registry_view=32"

reg add "HKCU\Software\Microsoft\Command Processor" ^
    /v "AutoRun" /t "REG_EXPAND_SZ" /f /reg:%registry_view% ^
    /d "@if exist """^%USERPROFILE^%\.cmd""" """^%USERPROFILE^%\.cmd""""
```
</details>

<details>
<summary>Explanation</summary>

From [Console Virtual Terminal Sequences](https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences): `ESC [ ? 12 l` disables text cursor blinking. The problem is then twofold:

- emit said sequence within a terminal
- make the operation persistent on startup

There are many ways of generating ESC within `cmd.exe`, but the cleanest method involves printing it with the [`prompt`](https://ss64.com/nt/prompt.html) internal command:

```batchfile
for /f "usebackq" %%E in (`echo(prompt $E^| "%ComSpec%"`) do set esc=%%~E
```

1. `prompt $E` sets the current prompt to an ESC character.
2. `echo` that command and pipe it to the default command interpreter (`cmd.exe`) that's stored in a global environment variable `ComSpec` to make it set a new prompt and output it. Why `echo` with an unclosed parenthesis? [For a multitude of good reasons.](https://www.dostips.com/forum/viewtopic.php?p=4554#p4554)
3. Execute the above in a subshell (within backticks, like other commonplace shells) and store the output in a `for /f` variable named `E`. As the contents of the command are going to be evaluated again, that `|` has to be escaped. The tilde is technically not necessary (it serves the purpose of removing surrounding double quotes), but is considered good practice. 
4. As that is going to run within a context of a batch script, double all the `%`s.

With ESC in a variable we're free to disable blinking permanently:

```batchfile
echo(%esc%[?12l %esc%[3A %esc%[3M %esc%[1A
```

In order:

1. Do what we came here for. This will spew some rubbish to the buffer. (todo control execution order)
2. Move the cursor up by three lines so that all of the crap printed is either below or on the same line.
3. Erase 3 lines from the buffer starting from the position we're on, cleaning up the screen.
4. Finally, move the cursor up by a line. This is not necessary, but it will cleanly separate prompt from startup information when the control returns.

From here we're off to making said changes persistent - with `cmd.exe`'s help (literally - `cmd /?`):

```
If /D was NOT specified on the command line, then when CMD.EXE starts, it
looks for the following REG_SZ/REG_EXPAND_SZ registry variables, and if
either or both are present, they are executed first.

    HKEY_LOCAL_MACHINE\Software\Microsoft\Command Processor\AutoRun

        and/or

    HKEY_CURRENT_USER\Software\Microsoft\Command Processor\AutoRun
```

Make the content of that registry variable execute a script of our choice if it exists:

```batchfile
set PROCESSOR_ARCHITECTURE | "%SystemRoot%\system32\find.exe" "64" >nul && (
    set "registry_view=64") || set "registry_view=32"

reg add "HKCU\Software\Microsoft\Command Processor" ^
    /v "AutoRun" /t "REG_EXPAND_SZ" /reg:%registry_view% ^
    /d "@if exist """^%USERPROFILE^%\.cmd""" """^%USERPROFILE^%\.cmd""""
```

Such file probably isn't present yet, so create a `.cmd` file in the current user's home directory:

```batchfile
type nul >"%USERPROFILE%\.cmd"
```

And populate it with the following:

```batchfile
@setlocal EnableDelayedExpansion EnableExtensions & set "CMD=!CMDCMDLINE!" 2>nul
@if /i not "!CMD!"=="!CMD:/=!" (goto :EOF) else @if not defined DEBUG (echo off)

for /f "usebackq" %%E in (`echo(prompt $E^| "%ComSpec%" /d /q "" 2^>nul`) do (
    echo(%%~E[?12l %%~E[3A %%~E[3M %%~E[1A
)
```

Take note that this script will run every time an instance of `cmd.exe` starts. That involves **all subshells**, including both the one we're getting an ESC character from and the one wrapping the whole command. To prevent that, we're creating a kind of a header guard in the first two lines - it will exit immediately if the currently executing command contains slashes (as subshells are basically `cmd /c <backtick_content>`). That doesn't involve the instance we're piping the prompt command to, thus a `/d` switch is added to prevent it from executing any startup scripts.

todo debugging machinery, explanation of delayed expansion and `cmd /q 2>nul`, exit label, `@` operator, sleep
</details>

<details>
<summary>Preview</summary>

https://github.com/microsoft/terminal/assets/7210216/e77b007f-5926-4127-87bb-0c721a39bf53

</details>

**Comment by GeekDuanLian** [comment]
For those using `bash`, add this to your `~/.bashrc`:
`printf '\e[?12l'`

**Comment by lost22git** [comment]
For those using WindowsPowerShell/pwsh, add this to your `$PROFILE`

```powershell
echo "`e[?12l"
```

**Comment by abhbh** [comment]
For Nushell users, add the following to your `config.nu`.

```
print "\e[?12l"
```

**Comment by densalo** [comment]
Σε windows 10 με windows terminal 1.20.11781.0 και PowerShell 7.4.4
δεν έχω καταφέρει to stop cursor blinking.

**Comment by atrme** [comment]
> On windows 10 with windows terminal 1.20.11781.0 and PowerShell 7.4.4 I have not been able to stop cursor blinking. I've tried all forms of _e[_ but it doesn't seem to do anything.

You may try this instead since "\`e" doesn't work on newer versions of Powershell

    $esc = char[0x1b]  
    Write-Host -n "${esc}[2 q"

**Comment by nenoganchev** [comment]
The workaround with `echo -e "\e[2 q"` is good enough for the time a terminal starts, I have it in my `~/.profile` and it's working. But Neovim [recently](https://github.com/neovim/neovim/pull/40234) started resetting the cursor upon exit in a way that now affects Windows Terminal as well. The end result is that my cursor is once again blinking after exiting `nvim` (which is at least on every `git commit`) and so it needs constant manual resetting back to non-blinking.

AFAICT, the only proper solution is to be able to set the blinking state in Windows Terminal, so that we have control over what a reset state looks like. There's only so much you can do with workarounds.

**Comment by stanio** [comment]
I have my shell prompt include `\e[2 q` so it gets reset on every prompt, but I'm seeing the blinking cursor after exiting the editor during an interactive git rebase, until completes – sometimes this is slow and is very annoying seeing the cursor blinking during that time.
